src/System.Private.CoreLib/shared/System/InvariantArray.cs

```diff
@@ -0,0 +1,181 @@
+        public Span<T> Span
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if (IsNull)
+                {
+                    return default;
+                }
+                else
+                {
+                    return new Span<T>(ref Unsafe.As<byte, T>(ref _array.GetRawSzArrayData()), _array.Length);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the backing array as a mutable <see cref="Span{T}"/>.
+        /// </summary>
+        /// <exception cref="NullReferenceException">
+        /// If <see cref="IsNull"/> is true.
+        /// </exception>
+        public Span<T> SpanNotNull
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            [StackTraceHidden]
+            get
+            {
+                T[] array = _array;
+
+                // Note: Accessing Length first avoids an unnecessary cmp instruction.
+
+                return new Span<T>(
+                    length: array.Length,
+                    ptr: ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData()));
+            }
+        }
+
+        /// <summary>
+        /// Reads a value from or writes a value to the backing array at the specified index.
+        /// The access uses normal array bounds checking.
+        /// </summary>
+        public T this[int index]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _array[index];
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            set
+            {
+                if (RuntimeHelpers.IsReference<T>())
+                {
+                    // Normally when a write takes place to a SzArray of reference type T,
+                    // the JIT emits a call to JIT_Stelem_Ref, which performs five actions:
+                    //
+                    // 1) check that the array isn't null,
+                    // 2) check that the index isn't out of bounds,
+                    // 3) check that the value to insert is the correct type,
+                    // 4) write the value to memory, then
+                    // 5) update the GC card table.
+                    //
+                    // One workaround is to get a `ref T` and to write directly to that,
+                    // but this causes the JIT to emit a call to JIT_CheckedWriteBarrier,
+                    // which performs three actions:
+                    //
+                    // 1) check that the address to write to is within the GC heap,
+                    // 2) write the value to memory, then
+                    // 3) update the GC card table.
+                    //
+                    // Since we're backed by a managed array (which must always be in the
+                    // GC heap), we can get the JIT to emit a call to the faster
+                    // JIT_WriteBarrier helper method, which only performs two actions:
```

Comment:
Are there still any bound-checks? I'm not aware of `JIT_WriteBarrier`, but `index` seems to be not validated according the description and the code in this PR.

Reply:
Yes, this line incurs a bounds check (unless the caller has already performed the bounds check and the JIT can elide it) because it's still a normal array access from the JIT's perspective:

```cs
Unsafe.As<InvariantArrayObjectWrapper[]>(_array)[index] = new InvariantArrayObjectWrapper(value);
```